ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs

```diff
@@ -48,13 +63,77 @@ instance (OuroborosTag p) => OuroborosTag (WithEBBs p) where
 
   preferCandidate       (WithEBBNodeConfig cfg) = preferCandidate       cfg
   compareCandidates     (WithEBBNodeConfig cfg) = compareCandidates     cfg
-  checkIsLeader         (WithEBBNodeConfig cfg) = checkIsLeader         cfg
-  rewindChainState      (WithEBBNodeConfig cfg) = rewindChainState      cfg
   protocolSecurityParam (WithEBBNodeConfig cfg) = protocolSecurityParam cfg
 
   applyChainState (WithEBBNodeConfig cfg) lv b cs = case
     eitherHeaderOrEbb cfg b of
-      Right hdr -> applyChainState @p cfg lv hdr cs
+      Right hdr ->
+          wrap <$> applyChainState @p cfg lv hdr (underlyingChainState cs)
+        where
+          wrap x = ChainStateWithEBBs
+            { firstNonEBBSlot      = Just $ case firstNonEBBSlot cs of
+                Just s  -> s
+                Nothing -> blockSlotOfHeader cfg [b] hdr
+            , underlyingChainState = x
+            }
       -- Currently the hash chain is maintained in the ledger state, so we just
       -- ignore the EBB for protocol concerns.
       Left _ -> return cs
+
+  checkIsLeader (WithEBBNodeConfig cfg) slot lv cs =
+    checkIsLeader cfg slot lv (underlyingChainState cs)
+
+  rewindChainState (WithEBBNodeConfig cfg) cs mSlot =
+      wrap <$> rewindChainState cfg (underlyingChainState cs) mSlot'
+    where
+      effectivelyRewindingToOrigin = case mSlot of
+        At slot
+          | Just s <- firstNonEBBSlot cs -> slot < s
+        _                                -> True
+
+      (mSlot', firstNonEBBSlot')
+         | effectivelyRewindingToOrigin = (Origin, Nothing)
+         | otherwise                    = (mSlot, firstNonEBBSlot cs)
+
+      wrap x = ChainStateWithEBBs
+        { firstNonEBBSlot      = firstNonEBBSlot'
+        , underlyingChainState = x
+        }
+
+-- | A wrapper around @'ChainState' p@ with a little extra bookeeping data.
+--
+data ChainStateWithEBBs p = ChainStateWithEBBs
```

Comment:
Done. The comment I wrote highlights the specificity of this PR. It only addressed the one failure mode we've witnessed: `PBFT` refuses to rewind to an empty chain unless `Origin` was specifically requested. Issue 925 is more general than that, but this PR doesn't handle the general case. For example `Praos` hasn't failed due to 925, but its comment sure sounds like it could -- note especially the parenthetical:

https://github.com/input-output-hk/ouroboros-network/blob/723d0ae621bfbc5b95b5813820badf90da22340f/ouroboros-consensus/src/Ouroboros/Consensus/Protocol/Praos.hs#L280-L291